java/src/main/native/OrtJniUtil.c

```diff
@@ -600,427 +588,473 @@ size_t copyPrimitiveArrayToJava(JNIEnv *jniEnv, ONNXTensorElementDataType onnxTy
-    } else {
-        // recurse through the dimensions
-        // Java arrays are objects until the final dimension
-        jobjectArray outputObjArr = (jobjectArray) output;
-        uint32_t dimLength = (*jniEnv)->GetArrayLength(jniEnv,outputObjArr);
-        size_t sizeConsumed = 0;
-        for (uint32_t i = 0; i < dimLength; i++) {
-            jarray childArr = (jarray) (*jniEnv)->GetObjectArrayElement(jniEnv,outputObjArr,i);
-            sizeConsumed += copyTensorToJava(jniEnv, onnxType, tensor + sizeConsumed, tensorSize - sizeConsumed, dimensionsRemaining - 1, childArr);
-            // Cleanup reference to childArr so it doesn't prevent GC.
-            (*jniEnv)->DeleteLocalRef(jniEnv,childArr);
-        }
-        return sizeConsumed;
+  if (dimensionsRemaining == 1) {
+    // write out 1d array of the respective primitive type
+    return copyPrimitiveArrayToJava(jniEnv, onnxType, tensor, output);
+  } else {
+    // recurse through the dimensions
+    // Java arrays are objects until the final dimension
+    jobjectArray outputObjArr = (jobjectArray) output;
+    int32_t dimLength = (*jniEnv)->GetArrayLength(jniEnv, outputObjArr);
+    int64_t sizeConsumed = 0;
+    for (int32_t i = 0; i < dimLength; i++) {
+      jarray childArr = (jarray) (*jniEnv)->GetObjectArrayElement(jniEnv, outputObjArr, i);
+      int64_t consumed = copyTensorToJava(jniEnv, onnxType, tensor + sizeConsumed, tensorSize - sizeConsumed, dimensionsRemaining - 1, childArr);
+      sizeConsumed += consumed;
+      // Cleanup reference to childArr so it doesn't prevent GC.
+      (*jniEnv)->DeleteLocalRef(jniEnv, childArr);
+      // If we failed to copy an array then break and return.
+      if (consumed == -1) {
+        return -1;
+      }
     }
+    return sizeConsumed;
+  }
 }
 
 jobject createStringFromStringTensor(JNIEnv *jniEnv, const OrtApi * api, OrtAllocator* allocator, OrtValue* tensor) {
-    // Get the buffer size needed
-    size_t totalStringLength;
-    checkOrtStatus(jniEnv,api,api->GetStringTensorDataLength(tensor,&totalStringLength));
-
-    // Create the character and offset buffers
-    char * characterBuffer;
-    checkOrtStatus(jniEnv,api,api->AllocatorAlloc(allocator,sizeof(char)*(totalStringLength+1),(void**)&characterBuffer));
-    size_t * offsets;
-    checkOrtStatus(jniEnv,api,api->AllocatorAlloc(allocator,sizeof(size_t),(void**)&offsets));
+  jobject tempString = NULL;
+  // Get the buffer size needed
+  size_t totalStringLength = 0;
+  OrtErrorCode code = checkOrtStatus(jniEnv, api, api->GetStringTensorDataLength(tensor, &totalStringLength));
+  if (code != ORT_OK) {
+    return NULL;
+  }
 
+  // Create the character and offset buffers, character is one larger to allow zero termination.
+  char * characterBuffer = NULL;
+  code = checkOrtStatus(jniEnv, api, api->AllocatorAlloc(allocator, sizeof(char)*(totalStringLength+1), (void**)&characterBuffer));
+  if (code != ORT_OK) {
+    return NULL;
```

Comment:
Just like in other cases, we can replace with malloc/free for simpler code and greater clarify. The allocator parameter is not really necessary since we always operate with CPU based memory.

Reply:
Sure.